### ISSUE TITLE
Conv Nets: A Modular Perspective

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10666]
Conv Nets: A Modular Perspective<br>
<br>
https://ift.tt/ryPQnUt<br>
<br>
<br>
<br>

<div><div><div><div>
<div>
<h2>Introduction</h2>
<p>In the last few years, deep neural networks have lead to breakthrough results on a variety of pattern recognition problems, such as computer vision and voice recognition. One of the essential components leading to these results has been a special kind of neural network called a <em>convolutional neural network</em>.</p>
<p>At its most basic, convolutional neural networks can be thought of as a kind of neural network that uses many identical copies of the same neuron. This allows the network to have lots of neurons and express computationally large models while keeping the number of actual parameters – the values describing how neurons behave – that need to be learned fairly small.</p>
<div>
<img src="https://colah.github.io/posts/2014-07-Conv-Nets-Modular/img/Conv2-9x5-Conv2Conv2.png"><p>
A 2D Convolutional Neural Network
</p>
</div>
<p>
</p>
<p>This trick of having multiple copies of the same neuron is roughly analogous to the abstraction of functions in mathematics and computer science. When programming, we write a function once and use it in many places – not writing the same code a hundred times in different places makes it faster to program, and results in fewer bugs. Similarly, a convolutional neural network can learn a neuron once and use it in many places, making it easier to learn the model and reducing error.</p>
<h2>Structure of Convolutional Neural Networks</h2>
<p>Suppose you want a neural network to look at audio samples and predict whether a human is speaking or not. Maybe you want to do more analysis if someone is speaking.</p>
<p>You get audio samples at different points in time. The samples are evenly spaced.</p>
<p>
</p>
<p>The simplest way to try and classify them with a neural network is to just connect them all to a fully-connected layer. There are a bunch of different neurons, and every input connects to every neuron.</p>
<p>
</p>
<p>A more sophisticated approach notices a kind of <em>symmetry</em> in the properties it’s useful to look for in the data. We care a lot about local properties of the data: What frequency of sounds are there around a given time? Are they increasing or decreasing? And so on.</p>
<p>We care about the same properties at all points in time. It’s useful to know the frequencies at the beginning, it’s useful to know the frequencies in the middle, and it’s also useful to know the frequencies at the end. Again, note that these are local properties, in that we only need to look at a small window of the audio sample in order to determine them.</p>
<p>So, we can create a group of neurons, <span>\(A\)</span>, that look at small time segments of our data. <span>\(A\)</span> looks at all such segments, computing certain <em>features</em>. Then, the output of this <em>convolutional layer</em> is fed into a fully-connected layer, <span>\(F\)</span>.</p>
<p>
</p>
<p>In the above example, <span>\(A\)</span> only looked at segments consisting of two points. This isn’t realistic. Usually, a convolution layer’s window would be much larger.</p>
<p>In the following example, <span>\(A\)</span> looks at 3 points. That isn’t realistic either – sadly, it’s tricky to visualize <span>\(A\)</span> connecting to lots of points.</p>
<p>
</p>
<p>One very nice property of convolutional layers is that they’re composable. You can feed the output of one convolutional layer into another. With each layer, the network can detect higher-level, more abstract features.</p>
<p>In the following example, we have a new group of neurons, <span>\(B\)</span>. <span>\(B\)</span> is used to create another convolutional layer stacked on top of the previous one.</p>
<p>
</p>
<p>Convolutional layers are often interweaved with pooling layers. In particular, there is a kind of layer called a max-pooling layer that is extremely popular.</p>
<p>Often, from a high level perspective, we don’t care about the precise point in time a feature is present. If a shift in frequency occurs slightly earlier or later, does it matter?</p>
<p>A max-pooling layer takes the maximum of features over small blocks of a previous layer. The output tells us if a feature was present in a region of the previous layer, but not precisely where.</p>
<p>Max-pooling layers kind of “zoom out”. They allow later convolutional layers to work on larger sections of the data, because a small patch after the pooling layer corresponds to a much larger patch before it. They also make us invariant to some very small transformations of the data.</p>
<p>
</p>
<p>In our previous examples, we’ve used 1-dimensional convolutional layers. However, convolutional layers can work on higher-dimensional data as well. In fact, the most famous successes of convolutional neural networks are applying 2D convolutional neural networks to recognizing images.</p>
<p>In a 2-dimensional convolutional layer, instead of looking at segments, <span>\(A\)</span> will now look at patches.</p>
<p>For each patch, <span>\(A\)</span> will compute features. For example, it might learn to detect the presence of an edge. Or it might learn to detect a texture. Or perhaps a contrast between two colors.</p>
<p>
</p>
<p>In the previous example, we fed the output of our convolutional layer into a fully-connected layer. But we can also compose two convolutional layers, as we did in the one dimensional case.</p>
<p>
</p>
<p>We can also do max pooling in two dimensions. Here, we take the maximum of features over a small patch.</p>
<p>What this really boils down to is that, when considering an entire image, we don’t care about the exact position of an edge, down to a pixel. It’s enough to know where it is to within a few pixels.</p>
<p>
</p>
<p>Three-dimensional convolutional networks are also sometimes used, for data like videos or volumetric data (eg. 3D medical scans). However, they are not very widely used, and much harder to visualize.</p>
<p>Now, we previously said that <span>\(A\)</span> was a group of neurons. We should be a bit more precise about this: what is <span>\(A\)</span> exactly?</p>
<p>In traditional convolutional layers, <span>\(A\)</span> is a bunch of neurons in parallel, that all get the same inputs and compute different features.</p>
<p>For example, in a 2-dimensional convolutional layer, one neuron might detect horizontal edges, another might detect vertical edges, and another might detect green-red color contrasts.</p>
<p>
</p>
<p>That said, in the recent paper ‘Network in Network’ (<a href="http://arxiv.org/abs/1312.4400">Lin <em>et al.</em> (2013)</a>), a new “Mlpconv” layer is proposed. In this model, <span>\(A\)</span> would have multiple layers of neurons, with the final layer outputting higher level features for the region. In the paper, the model achieves some very impressive results, setting new state of the art on a number of benchmark datasets.</p>
<p>
</p>
<p>That said, for the purposes of this post, we will focus on standard convolutional layers. There’s already enough for us to consider there!</p>
<h2>Results of Convolutional Neural Networks</h2>
<p>Earlier, we alluded to recent breakthroughs in computer vision using convolutional neural networks. Before we go on, I’d like to briefly discuss some of these results as motivation.</p>
<p>In 2012, Alex Krizhevsky, Ilya Sutskever, and Geoff Hinton blew existing image classification results out of the water (<a href="http://www.cs.toronto.edu/~fritz/absps/imagenet.pdf">Krizehvsky <em>et al.</em> (2012)</a>).</p>
<p>Their progress was the result of combining together a bunch of different pieces. They used GPUs to train a very large, deep, neural network. They used a new kind of neuron (ReLUs) and a new technique to reduce a problem called ‘overfitting’ (DropOut). They used a very large dataset with lots of image categories (<a href="http://www.image-net.org/">ImageNet</a>). And, of course, it was a convolutional neural network.</p>
<p>Their architecture, illustrated below, was very deep. It has 5 convolutional layers, with pooling interspersed, and three fully-connected layers. The early layers are split over the two GPUs.</p>
<p>
</p>
<p>They trained their network to classify images into a thousand different categories.</p>
<p>Randomly guessing, one would guess the correct answer 0.1% of the time. Krizhevsky, <em>et al.</em>’s model is able to give the right answer 63% of the time. Further, one of the top 5 answers it gives is right 85% of the time!</p>
<div>
<img src="https://colah.github.io/posts/2014-07-Conv-Nets-Modular/img/KSH-results.png"><div><p>
Top: 4 correctly classified examples. Bottom: 4 incorrectly classified examples. Each example has an image, followed by its label, followed by the top 5 guesses with probabilities. From <a href="http://www.cs.toronto.edu/~fritz/absps/imagenet.pdf">Krizehvsky <em>et al.</em> (2012)</a>.
</p></div>
</div>
<p>
</p>
<p>Even some of its errors seem pretty reasonable to me!</p>
<p>We can also examine what the first layer of the network learns to do.</p>
<p>Recall that the convolutional layers were split between the two GPUs. Information doesn’t go back and forth each layer, so the split sides are disconnected in a real way. It turns out that, every time the model is run, the two sides specialize.</p>
<div>
<img src="https://colah.github.io/posts/2014-07-Conv-Nets-Modular/img/KSH-filters.png"><div>
<p>
Filters learned by the first convolutional layer. The top half corresponds to the layer on one GPU, the bottom on the other. From </p>
<a href="http://www.cs.toronto.edu/~fritz/absps/imagenet.pdf">Krizehvsky <em>et al.</em> (2012)</a>
</div>
</div>
<p>
</p>
<p>Neurons in one side focus on black and white, learning to detect edges of different orientations and sizes. Neurons on the other side specialize on color and texture, detecting color contrasts and patterns. Remember that the neurons are <em>randomly</em> initialized. No human went and set them to be edge detectors, or to split in this way. It arose simply from training the network to classify images.</p>
<p>These remarkable results (and other exciting results around that time) were only the beginning. They were quickly followed by a lot of other work testing modified approaches and gradually improving the results, or applying them to other areas. And, in addition to the neural networks community, many in the computer vision community have adopted deep convolutional neural networks.</p>
<p>Convolutional neural networks are an essential tool in computer vision and modern pattern recognition.</p>
<h2>Formalizing Convolutional Neural Networks</h2>
<p>Consider a 1-dimensional convolutional layer with inputs <span>\(\{x_n\}\)</span> and outputs <span>\(\{y_n\}\)</span>:</p>
<p>
</p>
<p>It’s relatively easy to describe the outputs in terms of the inputs:</p>
<p><span>\[y_n = A(x_{n}, x_{n+1}, ...)\]</span></p>
<p>For example, in the above:</p>
<p><span>\[y_0 = A(x_0, x_1)\]</span> <span>\[y_1 = A(x_1, x_2)\]</span></p>
<p>Similarly, if we consider a 2-dimensional convolutional layer, with inputs <span>\(\{x_{n,m}\}\)</span> and outputs <span>\(\{y_{n,m}\}\)</span>:</p>
<p>
</p>
<p>We can, again, write down the outputs in terms of the inputs:</p>
<p><span>\[y_{n,m} = A\left(\begin{array}{ccc} x_{n,~m}, &amp; x_{n+1,~m},&amp; ...,~\\ x_{n,~m+1}, &amp; x_{n+1,~m+1}, &amp; ..., ~\\ &amp;...\\\end{array}\right)\]</span></p>
<p>For example:</p>
<p><span>\[y_{0,0} = A\left(\begin{array}{cc} x_{0,~0}, &amp; x_{1,~0},~\\ x_{0,~1}, &amp; x_{1,~1}~\\\end{array}\right)\]</span> <span>\[y_{1,0} = A\left(\begin{array}{cc} x_{1,~0}, &amp; x_{2,~0},~\\ x_{1,~1}, &amp; x_{2,~1}~\\\end{array}\right)\]</span></p>
<p>If one combines this with the equation for <span>\(A(x)\)</span>,</p>
<p><span>\[A(x) = \sigma(Wx + b)\]</span></p>
<p>one has everything they need to implement a convolutional neural network, at least in theory.</p>
<p>In practice, this is often not best way to think about convolutional neural networks. There is an alternative formulation, in terms of a mathematical operation called <em>convolution</em>, that is often more helpful.</p>
<p>The convolution operation is a powerful tool. In mathematics, it comes up in diverse contexts, ranging from the study of partial differential equations to probability theory. In part because of its role in PDEs, convolution is very important in the physical sciences. It also has an important role in many applied areas, like computer graphics and signal processing.</p>
<p>For us, convolution will provide a number of benefits. Firstly, it will allow us to create much more efficient implementations of convolutional layers than the naive perspective might suggest. Secondly, it will remove a lot of messiness from our formulation, handling all the bookkeeping presently showing up in the indexing of <span>\(x\)</span>s – the present formulation may not seem messy yet, but that’s only because we haven’t got into the tricky cases yet. Finally, convolution will give us a significantly different perspective for reasoning about convolutional layers.</p>
<blockquote>
<p>I admire the elegance of your method of computation; it must be nice to ride through these fields upon the horse of true mathematics while the like of us have to make our way laboriously on foot.  — Albert Einstein</p>
</blockquote>
<h2>Next Posts in this Series</h2>
<p><a href="https://colah.github.io/posts/2014-07-Understanding-Convolutions/"><strong>Read the next post!</strong></a></p>
<p>This post is part of a series on convolutional neural networks and their generalizations. The first two posts will be review for those familiar with deep learning, while later ones should be of interest to everyone. To get updates, subscribe to my <a href="https://colah.github.io/rss.xml">RSS feed</a>!</p>
<p>Please comment below or on the side. Pull requests can be made on <a href="https://github.com/colah/Conv-Nets-Series">github</a>.</p>
<h2>Acknowledgments</h2>
<p>I’m grateful to Eliana Lorch, Aaron Courville, and Sebastian Zany for their comments and support.</p>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via  <br>
<br>
January 23, 2024 at 11:00AM
